README.md
tv_grab_uk_atlas
----------------

Fetch TV and radio programme listings from "Atlas" data store and reformat into XMLTV format.


RATIONALE
---------

Most people in the UK use the excellent Radio Times ".dat" file feed to obtain TV and radio programme schedules for their PVR or Listings Viewer.

The Radio Times service was taken over by MetaBroadcast in December 2011 who continue to provide the feed.  However MetaBroadcast have stated it is their intention to terminate the .dat feed at some point.  Further they are no longer actively maintaining this feed and have publicly stated they will not be adding any new channels to it.  So as new channels become available, there is no schedule info available for them.  

At the time of writing (Sep. 2013) an increasing number of channels are not available in the .dat feed (e.g. Drama, Movie Mix, BT Sport, etc.)

In place of the .dat feed MetaBroadcast would like people to use their "Atlas" database via its API.  This module does just that by building on the "XMLTV" package.



OPERATION
---------

Drop-in module for the XMLTV package, run as a command-line script.  Output to a disc file in XMLTV format.

Includes a CGI script to run via a web browser.



PRE-REQUISITES
--------------

Assumes you already have a working copy of "XMLTV".

Optional: web server if you want to use the web browser front-end.



COMPATABILITY
-------------

Tested with XMLTV 0.5.63, Perl v5.8.8 and Apache 2.2.3

(Tech note: many web hosts provide only 5.8.8 so no features are included which require a more recent version of Perl.)



VARIANCE
--------

The Atlas data store does not include "RadioTimes Choices" nor film ratings.



LINKS
-----

"Atlas" : http://atlas.metabroadcast.com/

XMLTV DTD : http://xmltv.cvs.sourceforge.net/viewvc/xmltv/xmltv/xmltv.dtd

XMLTV : http://wiki.xmltv.org/index.php/Main_Page

"Changes to the Radio Times XML TV service" : http://www.radiotimes.com/blog/2011-12-12/changes-to-the-radio-times-xml-tv-service

"providing the RadioTimes XMLTV feed from atlas" : http://metabroadcast.com/blog/providing-the-radiotimes-xmltv-feed-from-atlas


CHANGELOG
1.27   28-Aug-2014  Don't error if user.map file not found
1.26   15-Aug-2014  Kill bad incoming data for series/ep numbers
1.25    1-Aug-2014  Fix Perl error when filtering empty programmes hash (credit: ta264) (minor)
                    Report error returned by Atlas when status 400 received (minor)
1.24   12-Jul-2014  Remove audio described (it is different to deaf signed) (minor)
1.23   28-May-2014  Fix encoding of Irish channel names in xml output (minor)
1.22   22-May-2014  Add duplicate programme deletion / clump creation
1.21   14-Mar-2014  Update instructions on obtaining your Atlas API key
1.20    7-Mar-2014  Don't change order of Film category when present as an Atlas genre (minor)
1.19    6-Mar-2014  Output Atlas genres before PA genres, and in the same order as they arrive
1.18   23-Jan-2014  Fix an issue with character encodings (minor)
1.17   21-Jan-2014  Enhance POD information
                    Use new --info option
                    Fix 'premiere' flag
1.16    9-Jan-2014  Disable widescreen, high-def and stereo tags (c.f. v1.14)
1.15    9-Jan-2014  Add test for XMLTV version
1.14   20-Dec-2013  Changes to attributes on website
1.13   18-Dec-2013  Add support for lineups capability (--list-lineups, --get-lineup)
                    Option to use 'lineup' code during configure
1.12   14-Dec-2013  Fix --list-channels so it uses the conf hash
1.11    5-Dec-2013  Minor mod to config validation
1.10    5-Dec-2013  Changed handling of map files
1.9     5-Dec-2013  Move apikey from separate file to --configure stage
1.8     4-Dec-2013  Improve error message when API Key missing
1.7    26-Nov-2013  Improve --help handling
1.6    23-Oct-2013  Make "prevent overlaps" the default when fetching programmes
1.5    20-Oct-2013  Fix DST issue with programmes fetched from source.
                    Fix ValidateGrabber reported failure due to programme overlap
1.4    19-Oct-2013  Fix error in 'platform' selection
1.3    19-Oct-2013  Fix wrong directory when using $ENV{XMLTV_SUPPLEMENT}
1.2    18-Oct-2013  Tidy up some code
1.1    17-Oct-2013  Rev no. reset to 1.1
# pre-SourceForge --v
1.013  19-Oct-2013  Fix error in --configure
1.012  17-Oct-2013  Add option to change publisher
1.011  14-Oct-2013  Add 'episode_num' tag with the Atlas ids for brand, series and item
1.010  12-Oct-2013  Additional person roles
                    Fix for platforms with no regions (e.g. Saorview)
                    Modified method for determining programme's Title / Sub-title
                    Add 'role' attribute to actors
1.009  3-Oct-2013   Map genre to multiple out values
                    Allow genre removal
1.008  28-Sep-2013  Fix --list-channels
                    Add <icon> to channel details
1.007  23-Sep-2013  Fix channel's <display-name>
--     18-Sep-2013  Validate the generated xml with xmllint
1.006  12-Sep-2013  Ignore comments in 'map' files
1.005  11-Sep-2013  Bug fix in timezone display
1.004  11-Sep-2013  Handle HTTP failures
1.003  11-Sep-2013  Added --dst param
                    Simplified 'from'/'to' code 
                    Allow the channel specified to "--channel" to be a 'mapped' channel name 
1.002  11-Sep-2013  Map 'Film' genre to alternative name if added by 'type' attribute
                    Fixed bug in 'to' time calculation for --hours with --offset
1.001  10-Sep-2013  Fix grabbing of "premiere"
1.000  10-Sep-2013  Initial release



data/getchannels.pl
#!/usr/bin/perl -w
#
# Get all the platforms and regions and channels and their mapping from Atlas and create 'map' files.
# (Also try to auto-map against the RT grabber's channel_ids file)
#
#
# Copyright G. Westcott - September 2013
#
# This code is distributed under the GNU General Public License v2 (GPLv2) .
#
# 

use strict;
use warnings;
use constant { true => 1, false => 0 };
use Data::Dumper;

use File::Path;
use POSIX qw(strftime);
use DateTime;
use Date::Parse;
use Encode;
use URI::Escape;

use JSON::PP;
use HTTP::Cookies;
use LWP::UserAgent;
my $lwp = initialise_ua();

my $debug = '1';

# ------------------------------------------------------------------------------------------------------------------------------------- #
# Let's play nice and use a short-term cache to reduce load on Atlas site
#
use HTTP::Cache::Transparent;
HTTP::Cache::Transparent::init( { 
    BasePath => '/tmp/cache/',
    NoUpdate => 60*60,			# cache time in seconds
		MaxAge => 4,						# flush time in hours
    Verbose => $debug,
} );


# ------------------------------------------------------------------------------------------------------------------------------------- #
# Get the channels from Atlas
my $ROOT_URL = 'http://atlas.metabroadcast.com/3.0/';
my $platforms = ();
my $channels = ();

get_channels();


# ------------------------------------------------------------------------------------------------------------------------------------- #
exit(0);


# #############################################################################
# # THE MEAT #####################################################################
# ------------------------------------------------------------------------------------------------------------------------------------- #

sub get_channels {

		fetch_platforms();
		#print Dumper($platforms);
		print_platforms();
		fetch_channels();
		#print Dumper($channels);exit;
		print_channels();
		print_channels_RT();
		
		return;
}



sub fetch_platforms {
		# Fetch Atlas' channel_groups
		
		#		http://atlas.metabroadcast.com/3.0/channel_groups.json?type=platform
		my $url = $ROOT_URL.'channel_groups.json?type=platform';
		#print $url ."\n";

		# Fetch the page
		my $res = $lwp->get( $url );
		
		if ($res->is_success) {
				#print $res->content;
				
				# Extract the available platforms
				my $data = JSON::PP->new()->utf8(0)->decode($res->content);
				$res = undef;

				my $channel_group = $data->{'channel_groups'};
				foreach (@$channel_group) {
						my %group = %$_;
						next unless ($group{'type'} eq 'platform');
				
						my %platform = ();
				
						$platform{'id'} 				= $group{'id'};
						$platform{'title'} 			= $group{'title'};
						$platform{'uri'} 				= $group{'uri'};
						#$platform{'available_countries'} 	= $group{'available_countries'};
						$platform{'countries'} = '';
						foreach my $country (@{$group{'available_countries'}}) {
							$platform{'countries'} .= $country . ($platform{'countries'} ne '' ? ' ' : '');
						}
						
						$platform{'regions'} = ();
						foreach my $region (@{$group{'regions'}}) {
							push @{$platform{'regions'}},  { 'id' => $region->{'id'}, 'title' =>  $region->{'title'} };
						}
						
						push @{$platforms}, \%platform;
				}
				
		} else {
				print $res->status_line . "\n";
		}
	
		return;
}

sub print_platforms {
		# Write a file of all the Atlas platforms with  id==title  ("platforms.txt")
		# Write a file of all the Atlas regions for each platform with  id==title   ("regions_xxxx.txt")
		#
		my $f = 'platforms.txt';
		open OUT, "> $f"  or die "Failed to open $f for writing";
		printf OUT '# PLATFORMS'."\n";
		
		foreach (@{$platforms}) {
				my %p = %$_;
				
				#printf '%s  %s '."\n", $p{'id'}, $p{'title'} ;
				printf OUT '%s==%s '."\n", $p{'id'}, $p{'title'} ;
				
				#my $f2 = 'regions_'. $p{'id'} .'.txt';
				( my $x = $p{'title'} ) =~ s/\s/_/;
				my $f2 = 'regions_'. $x .'.txt';			
				open OUT2, "> $f2"  or die "Failed to open $f2 for writing";
				printf OUT2 '# REGIONS for '.$p{'title'}."\n";
		
				foreach (@{$p{'regions'}}) {
						my %r = %$_;
						
						#printf '%s  %s '."\n", $r{'id'}, $r{'title'} ;
						printf OUT2 '%s==%s '."\n", $r{'id'}, $r{'title'} ;
				}
				
				close OUT2;
				
		}
		
		close OUT;
}


sub fetch_channels {
		# Fetch Atlas' channels for each region
		
		foreach (@$platforms) {
				my %p = %$_;
				
				foreach (@{$p{'regions'}}) {
						my %r = %$_;
						
						my $regioncode = $r{'id'};
						
						#		http://atlas.metabroadcast.com/3.0/channel_groups/cbhN.json?annotations=channels 
						#
						my $url = $ROOT_URL.'channel_groups/'.$regioncode.'.json?annotations=channels';
						#print $url ."\n" ;

						my %regionchannels = ();

						# Fetch the page
						my $res = $lwp->get( $url );
						
						if ($res->is_success) {
								#print $res->content;
								
								# Extract the available channels
								my $data = JSON::PP->new()->utf8(0)->decode($res->content);
								$res = undef;

								my $chans = $data->{'channel_groups'}[0]->{'channels'};
								foreach (@$chans) {
										my %chan = %$_;
										next unless ($chan{'channel'}->{'type'} eq 'channel');
								
										my %channel = ();
								
										$channel{'num'} 				= $chan{'channel_number'};
										$channel{'id'} 					= $chan{'channel'}->{'id'};
										$channel{'title'} 			= $chan{'channel'}->{'title'};
										$channel{'image'} 			= $chan{'channel'}->{'image'};
										$channel{'media_type'} 	= $chan{'channel'}->{'media_type'};		# 'video' 'audio'
										
										foreach (@{$chan{'channel'}->{'aliases'}}) {
												my $alias = $_;							
												#     'aliases' => [
												#                   'http://pressassociation.com/channels/1459',
												#                   'http://xmltv.radiotimes.com/channels/2569',
												#                   'supercasino',
												#                   'http://atlas.metabroadcast.com/4.0/channels/hmcb'
												#                 ],
												if ( $alias =~ m%http://xmltv.radiotimes.com/channels/(\d*)% ) {
													$channel{'rt_chan'} = $1;
													next;
												}
												if ( $alias =~ m%http://pressassociation.com/channels/(\d*)% ) {
													$channel{'pa_chan'} = $1;
													next;
												}
										}
												 
										push @{$regionchannels{$regioncode}}, \%channel;
								}
								
						} else {
								print $res->status_line . "\n";
						}
						
						push @{$channels}, \%regionchannels;					
				}
		}
	
		return;
}


sub print_channels {
		# Write a map file of all the Atlas channels for each region with  map==id==num   ("map_xxxx.txt")
		#
		foreach (@{$channels}) {
				my %r = %$_;
							
				foreach my $key (keys %r) {
						my $regioncode = $key;
						
						my $f = 'map_'. $regioncode .'.txt';			
						open OUT, "> $f"  or die "Failed to open $f for writing";
						printf OUT '# CHANNELS for '.$regioncode."\n";

						foreach (@{$r{$key}}) {
							my %c = %$_;
							# print Dumper(\%c);exit;
							
							printf OUT 'map==%s==%s '."\n", $c{'id'}, $c{'num'};
						}
							
						close OUT;
			}
				
		}
}


sub print_channels_RT {
		# Write a map file of all the Atlas channels for each region with  map==id==RT_id   ("map_xxxx_RT.txt")
		#
		# 'RT_id' is the channel_id (RFC2838 compliant) as defined in the channel_ids file from the tv_grab_uk_rt grabber
		#
		
		# Load the RT channel_ids
		my %rt_ids = ();
		#		
		my $fn = 'rt/channel_ids';
		my $fhok = open my $fh, '<', $fn or die("Cannot open file $fn");
		if ($fhok) {
			while (my $line = <$fh>) { 
				chomp $line;
				chop($line) if ($line =~ m/\r$/);
				next if $line =~ /^#/;
				my ($channel_id, $rt_id, $name, $icon, $offset, $txhours, $res) = split(/\|/, $line);
				#print "$channel_id, $rt_id, $name, $icon, $offset, $txhours, $res";
				
				$rt_ids{$rt_id} = $channel_id;
			}
			close $fh;
		}
		#print Dumper (\%rt_ids);
		
		
		foreach (@{$channels}) {
				my %r = %$_;
							
				foreach my $key (keys %r) {
						my $regioncode = $key;
						
						my $f = 'map_'. $regioncode .'_RT.txt';			
						open OUT, "> $f"  or die "Failed to open $f for writing";
						printf OUT '# CHANNELS for '.$regioncode."\n";
						printf OUT '# This is an attempt to auto-match the Atlas data against the '."\n";
						printf OUT '# RT grabber\'s channel_ids file'."\n";
						printf OUT '# '."\n";

						foreach (@{$r{$key}}) {
							my %c = %$_;
							#print Dumper(\%c);exit;
							
							my $rt_id = '';
							$rt_id = $rt_ids{$c{'rt_chan'}} if defined $c{'rt_chan'};
							if (!defined $rt_id) {
								print $c{'num'} . ' has no RT channel'."\n";
								$rt_id = '';
							}
							printf OUT "map==%s==%s   \t\t\t\t# %s (%s) - %s"."\n", $c{'id'}, $rt_id, $c{'num'}, (defined $c{'rt_chan'}?$c{'rt_chan'}:''), $c{'title'};
						}
							
						close OUT;
			}
				
		}
}


# #############################################################################
# # THE VEG ######################################################################
# ------------------------------------------------------------------------------------------------------------------------------------- #

sub initialise_ua {
		my $cookies = HTTP::Cookies->new;
		#my $ua = LWP::UserAgent->new(keep_alive => 1);
		my $ua = LWP::UserAgent->new;
		# Cookies
		$ua->cookie_jar($cookies);
		# Define user agent type
		$ua->agent('Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.0; Trident/5.0');
		# Define timouts
		$ua->timeout(240);
		# Use proxy if set in http_proxy etc.
		$ua->env_proxy;
		
		return $ua;
}


# #############################################################################

__END__


code/cgi-bin/getatlas.pl
#!/usr/bin/perl -w
#
# params are 'offset' and 'hours'
#				or  'offset' and 'days'    in which case the offset is 'days' also (otherwise it's 'hours')
#				or  'date'			- fetch just this day
#
# optional params:
#			channel		- Atlas channel id to fetch  (if not specified then grabber .conf file is used as normal)
#			dst				- Add an extra hour(s) to the schedule fetched (default '1'  ('0' if no 'dst' param specficied))
#
#

# You may need to set $HOME if not same as command profile
#$ENV{'HOME'} = '?????';

use strict;
use warnings;
use Data::Dumper;
use CGI; 
use CGI::Carp qw(fatalsToBrowser);

my $query = CGI->new;

# Fetch the query string params
my $offset 	= $query->param('offset');			# offset from now to start fetch (hours/days)
my $hours 	= $query->param('hours');				# hours to fetch
my $days 		= $query->param('days');				# days to fetch
my $channel = $query->param('channel');			# channel id or label
my $date 		= $query->param('date');				# YYYYMMDD
my $dst 		= $query->param('dst');					# (no value)
my $hasdst	= (defined $query->param('dst') ? 1 : 0);


# Validate the params
if ( ($hours && $days) ||
		 ($date && ($offset || $hours || $days)) ) {
	
		if (0) {	
			print <<END_OF_HTML;
Status: 500 Invalid Parameters
Content-type: text/html

<HTML>
<HEAD><TITLE>500 Invalid Parameters</TITLE></HEAD>
<BODY>
  <H1>Error</H1>
  <P>Invalid Parameters</P>
</BODY>
</HTML>
END_OF_HTML
		}

		if (1) {	
			print <<END_OF_XML;
Status: 500 Invalid Parameters
Content-type: text/xml

<?xml version="1.0" encoding="UTF-8"?>
END_OF_XML
		}
		
		exit;
}


my $action = '';
if ($hours) {
	$action = "--hours $hours " . ($offset ? "--offset $offset" : '');
} elsif ($days) {
	$action = "--days $days " . ($offset ? "--offset $offset" : '');
} elsif ($date) {
	$action = "--date $date";
} else {
	$action = "--days 1";
}

$channel = "--channel $channel" if $channel;
$dst 		 = "--dst" 							if $hasdst;



# Must send HTTP Content-Type header
#print CGI->header('text/xml');
# ^^ caused errors:     Use of uninitialized value in string ne at (eval 3) line 29.
#


# debug
#print "Content-type: text/xml"."\n\n";
#print "Params: <br /> $offset <br /> $hours <br /> $action <br />";exit(0);


# run the grabber (& capture the output)
system("perl $ENV{'HOME'}/tv_grab_uk_atlas.pl --quiet $action $channel $dst 1>/tmp/tv_grab_uk_atlas.stdout 2>/tmp/tv_grab_uk_atlas.stderr ");

my $result = '';
open(my $fh, '<', '/tmp/tv_grab_uk_atlas.stderr')
    or die "Unable to open file, $!";	
while(<$fh>) { 
   $result = $_ ;
   last;
}
close($fh)
    or warn "Unable to close the file handle: $!";

if ( $result =~ /^Status:/ ) {
	# grabber threw an error
	# assume it's a valid HTTP "Status: "
	print $result;
}
else {
	# validate the xml 
	system(' xmllint --noout --dtdvalid http://supplement.xmltv.org/xmltv.dtd /tmp/tv_grab_uk_atlas.stdout ');
	if ($? != 0) {
			print "Status: 500 Invalid XML \n\n";
			exit;
	}
}

# append the xml file (which is blank when an error occurs)
print "Content-type: text/xml"."\n\n";
system("cat /tmp/tv_grab_uk_atlas.stdout");
		

exit(0);

__END__


----------------------------------------------------------------------------------
# e.g. when error
stderr = 
---------------------------------------------------
Status: 400 Bad Request
---------------------------------------------------

stdout =
---------------------------------------------------
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE tv SYSTEM "xmltv.dtd">

<tv generator-info-name="tv_grab_uk_atlas" generator-info-url="http://atlas.metabroadcast.com/3.0/">
  <channel id="xxxx">
    <display-name lang="en">xxxx</display-name>
  </channel>
</tv>
---------------------------------------------------

return = 
---------------------------------------------------
HTTP/1.1 400 Bad Request
Content-type: text/xml

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE tv SYSTEM "xmltv.dtd">

<tv generator-info-name="tv_grab_uk_atlas" generator-info-url="http://atlas.metabroadcast.com/3.0/">
  <channel id="xxxx">
    <display-name lang="en">xxxx</display-name>
  </channel>
</tv>
---------------------------------------------------
